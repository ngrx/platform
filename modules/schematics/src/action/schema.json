{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxAction",
  "title": "NgRx Action Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the action.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the action?"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "skipTests": {
      "type": "boolean",
      "description": "When true, does not create test files.",
      "default": false
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "group": {
      "type": "boolean",
      "default": false,
      "description": "Group actions file within 'actions' folder",
      "aliases": ["g"]
    },
    "api": {
      "type": "boolean",
      "default": false,
      "description": "Specifies if api success and failure actions should be generated.",
      "aliases": ["a"],
      "x-prompt": "Should we generate success and failure actions?"
    },
    "creators": {
      "type": "boolean",
      "default": true,
      "description": "Specifies whether to use creator functions for handling actions and reducers.",
      "aliases": ["c"],
      "x-prompt": "Do you want to use the create function?"
    }
  },
  "required": []
}
