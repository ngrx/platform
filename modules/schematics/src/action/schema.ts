export interface Schema {
  /**
   * The name of the component.
   */

  name: string;
  /**
   * The path to create the component.
   */

  path?: string;
  /**
   * The name of the project.
   */
  project?: string;
  /**
   * Specifies if a spec file is generated.
   */
  spec?: boolean;
  /**
   * Flag to indicate if a dir is created.
   */

  flat?: boolean;
  /**
   * Group actions file within 'actions' folder
   */

  group?: boolean;
}
