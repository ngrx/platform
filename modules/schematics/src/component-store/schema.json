{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxComponentStore",
  "title": "NgRx Component Store Options Schema",
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the component store.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the component store?"
    },
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component store.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "flat": {
      "type": "boolean",
      "default": true,
      "description": "Flag to indicate if a dir is created."
    },
    "skipTests": {
      "type": "boolean",
      "default": false,
      "description": "When true, does not create test files."
    },
    "component": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring component.",
      "alias": "m",
      "subtype": "filepath",
      "x-prompt": "To which component (path) should the component store be provided in?"
    },
    "module": {
      "type": "string",
      "default": "",
      "description": "Allows specification of the declaring module.",
      "alias": "m",
      "subtype": "filepath",
      "x-prompt": "To which module (path) should the component store be provided in?"
    }
  },
  "required": []
}
