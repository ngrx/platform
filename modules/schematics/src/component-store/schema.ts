export interface Schema {
  /**
   * The name of the component store.
   */
  name: string;

  /**
   * The path to create the component store.
   */
  path?: string;

  /**
   * The name of the project.
   */
  project?: string;

  /**
   * Flag to indicate if a dir is created.
   */
  flat?: boolean;

  /**
   * When true, does not create test files.
   */
  skipTests?: boolean;

  /**
   * Allows specification of the declaring component.
   */
  component?: string;

  /**
   * Allows specification of the declaring module.
   */
  module?: string;
}
