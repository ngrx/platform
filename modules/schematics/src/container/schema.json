{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxContainer",
  "title": "NgRx Container Options Schema",
  "type": "object",
  "properties": {
    "path": {
      "type": "string",
      "format": "path",
      "description": "The path to create the component.",
      "visible": false
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "aliases": ["p"]
    },
    "name": {
      "type": "string",
      "description": "The name of the component.",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What should be the name of the container component?"
    },
    "inlineStyle": {
      "description": "Specifies if the style will be in the ts file.",
      "type": "boolean",
      "alias": "s"
    },
    "inlineTemplate": {
      "description": "Specifies if the template will be in the ts file.",
      "type": "boolean",
      "alias": "t"
    },
    "viewEncapsulation": {
      "description": "Specifies the view encapsulation strategy.",
      "enum": ["Emulated", "Native", "None"],
      "type": "string",
      "alias": "v"
    },
    "changeDetection": {
      "description": "Specifies the change detection strategy.",
      "enum": ["Default", "OnPush"],
      "type": "string",
      "alias": "c"
    },
    "prefix": {
      "type": "string",
      "format": "html-selector",
      "description": "The prefix to apply to generated selectors.",
      "alias": "p"
    },
    "style": {
      "description": "The file extension or preprocessor to use for style files.",
      "type": "string"
    },
    "skipTests": {
      "type": "boolean",
      "description": "When true, does not create test files."
    },
    "flat": {
      "type": "boolean",
      "description": "Flag to indicate if a dir is created."
    },
    "skipImport": {
      "type": "boolean",
      "description": "Flag to skip the module import."
    },
    "selector": {
      "type": "string",
      "format": "html-selector",
      "description": "The selector to use for the component."
    },
    "module": {
      "type": "string",
      "description": "Allows specification of the declaring module.",
      "alias": "m"
    },
    "export": {
      "type": "boolean",
      "description": "Specifies if declaring module exports the component."
    },
    "state": {
      "type": "string",
      "description": "Specifies the path to the state exports."
    },
    "stateInterface": {
      "type": "string",
      "default": "State",
      "description": "Specifies the interface for the state."
    },
    "testDepth": {
      "description": "Specifies whether to create a unit test or an integration test.",
      "enum": ["unit", "integration"],
      "type": "string",
      "default": "integration"
    }
  },
  "required": []
}
