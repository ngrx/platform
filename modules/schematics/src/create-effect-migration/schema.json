{
  "$schema": "http://json-schema.org/schema",
  "id": "SchematicsNgRxCreateEffectMigration",
  "title": "NgRx Effect Migration from @Effect to createEffect",
  "type": "object",
  "properties": {},
  "required": []
}
