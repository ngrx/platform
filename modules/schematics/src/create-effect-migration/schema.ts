// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Schema {}
