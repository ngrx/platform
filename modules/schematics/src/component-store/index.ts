import {
  apply,
  applyTemplates,
  branchAndMerge,
  chain,
  filter,
  mergeWith,
  move,
  noop,
  Rule,
  SchematicContext,
  SchematicsException,
  Tree,
  url,
} from '@angular-devkit/schematics';
import * as ts from 'typescript';
import { Schema as ComponentStoreOptions } from './schema';
import {
  addProviderToComponent,
  addProviderToModule,
  buildRelativePath,
  findComponentFromOptions,
  findModuleFromOptions,
  getProjectPath,
  InsertChange,
  parseName,
  stringUtils,
  insertImport,
} from '@ngrx/schematics/schematics-core';

interface AddProviderContext {
  componentStoreRelativePath: string;
  componentStoreName: string;
}

function createProvidingContext(
  options: Pick<ComponentStoreOptions, 'name' | 'flat' | 'path'>,
  providingPath: string
): AddProviderContext {
  const componentStoreName = `${stringUtils.classify(`${options.name}Store`)}`;
  const componentStorePath =
    `/${options.path}/` +
    (options.flat ? '' : stringUtils.dasherize(options.name) + '/') +
    stringUtils.dasherize(options.name) +
    '.store';
  const componentStoreRelativePath = buildRelativePath(
    providingPath,
    componentStorePath
  );

  return {
    componentStoreRelativePath,
    componentStoreName,
  };
}

/**
 * Add component store to NgModule
 */
export function addComponentStoreProviderToNgModule(
  options: ComponentStoreOptions
): Rule {
  return (host: Tree) => {
    if (!options.module) {
      return host;
    }

    const modulePath = options.module;
    if (!host.exists(options.module)) {
      throw new Error(`Specified module path ${modulePath} does not exist`);
    }

    const text = host.read(modulePath);
    if (text === null) {
      throw new SchematicsException(`File ${modulePath} does not exist.`);
    }
    const sourceText = text.toString('utf-8');

    const source = ts.createSourceFile(
      modulePath,
      sourceText,
      ts.ScriptTarget.Latest,
      true
    );

    const context = createProvidingContext(options, options.module);

    const componentStore = insertImport(
      source,
      modulePath,
      context.componentStoreName,
      context.componentStoreRelativePath
    );
    const [storeNgModuleProvider] = addProviderToModule(
      source,
      modulePath,
      context.componentStoreName,
      context.componentStoreRelativePath
    );
    const changes = [componentStore, storeNgModuleProvider];
    const recorder = host.beginUpdate(modulePath);
    for (const change of changes) {
      if (change instanceof InsertChange) {
        recorder.insertLeft(change.pos, change.toAdd);
      }
    }
    host.commitUpdate(recorder);

    return host;
  };
}

/**
 * Add component store to Component
 */
export function addComponentStoreProviderToComponent(
  options: ComponentStoreOptions
): Rule {
  return (host: Tree) => {
    if (!options.component) {
      return host;
    }

    const componentPath = options.component;
    if (!host.exists(options.component)) {
      throw new Error(
        `Specified component path ${componentPath} does not exist`
      );
    }

    const text = host.read(componentPath);
    if (text === null) {
      throw new SchematicsException(`File ${componentPath} does not exist.`);
    }
    const sourceText = text.toString('utf-8');

    const source = ts.createSourceFile(
      componentPath,
      sourceText,
      ts.ScriptTarget.Latest,
      true
    );

    const context = createProvidingContext(options, options.component);

    const componentStore = insertImport(
      source,
      componentPath,
      context.componentStoreName,
      context.componentStoreRelativePath
    );
    const [storeNgModuleProvider] = addProviderToComponent(
      source,
      componentPath,
      context.componentStoreName,
      context.componentStoreRelativePath
    );
    const changes = [componentStore, storeNgModuleProvider];
    const recorder = host.beginUpdate(componentPath);
    for (const change of changes) {
      if (change instanceof InsertChange) {
        recorder.insertLeft(change.pos, change.toAdd);
      }
    }
    host.commitUpdate(recorder);

    return host;
  };
}

export default function (options: ComponentStoreOptions): Rule {
  return (host: Tree, context: SchematicContext) => {
    options.path = getProjectPath(host, options);

    if (options.module) {
      options.module = findModuleFromOptions(host, options);
    }

    if (options.component) {
      options.component = findComponentFromOptions(host, options);
    }

    const parsedPath = parseName(options.path, options.name);
    options.name = parsedPath.name;
    options.path = parsedPath.path;

    const templateSource = apply(url('./files'), [
      options.skipTests
        ? filter((path) => !path.endsWith('.spec.ts.template'))
        : noop(),
      applyTemplates({
        ...stringUtils,
        'if-flat': (s: string) => (options.flat ? '' : s),
        ...options,
      }),
      move(parsedPath.path),
    ]);

    return chain([
      branchAndMerge(
        chain([
          addComponentStoreProviderToNgModule(options),
          addComponentStoreProviderToComponent(options),
          mergeWith(templateSource),
        ])
      ),
    ])(host, context);
  };
}
