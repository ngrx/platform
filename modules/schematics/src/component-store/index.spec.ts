import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import * as path from 'path';
import { Schema as ComponentStoreOptions } from '../component-store/schema';
import {
  createWorkspace,
  defaultAppOptions,
  defaultWorkspaceOptions,
  getTestProjectPath,
} from '@ngrx/schematics-core/testing';

describe('component-store', () => {
  const schematicRunner = new SchematicTestRunner(
    '@ngrx/schematics',
    path.join(__dirname, '../../collection.json')
  );

  const defaultOptions: ComponentStoreOptions = {
    name: 'foo',
    project: 'bar',
    module: undefined,
    flat: false,
  };

  const projectPath = getTestProjectPath();

  let appTree: UnitTestTree;

  beforeEach(async () => {
    appTree = await createWorkspace(schematicRunner, appTree);
  });

  it('should create an component store to specified project if provided', async () => {
    const options = {
      ...defaultOptions,
      project: 'baz',
    };

    const specifiedProjectPath = getTestProjectPath(defaultWorkspaceOptions, {
      ...defaultAppOptions,
      name: 'baz',
    });

    const tree = await schematicRunner
      .runSchematicAsync('cs', options, appTree)
      .toPromise();
    const files = tree.files;
    expect(
      files.indexOf(`${specifiedProjectPath}/src/lib/foo/foo.store.spec.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${specifiedProjectPath}/src/lib/foo/foo.store.ts`)
    ).toBeGreaterThanOrEqual(0);
  });

  it('should create an component store with a spec file', async () => {
    const options = { ...defaultOptions };

    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const files = tree.files;
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.store.spec.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.store.ts`)
    ).toBeGreaterThanOrEqual(0);
  });

  it('should not be provided by default', async () => {
    const options = { ...defaultOptions };

    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).not.toMatch(/import { FooStore } from '.\/foo\/foo.store'/);
  });

  it('should import into a specified module', async () => {
    const options = { ...defaultOptions, module: 'app.module.ts' };

    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);
    expect(content).toMatch(/import { FooStore } from '.\/foo\/foo.store'/);
  });

  it('should fail if specified module does not exist', async () => {
    const options = {
      ...defaultOptions,
      module: `${projectPath}/src/app/app.moduleXXX.ts`,
    };
    let thrownError: Error | null = null;
    try {
      await schematicRunner.runSchematicAsync(
        'component-store',
        options,
        appTree
      );
    } catch (err) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should respect the skipTests flag', async () => {
    const options = { ...defaultOptions, skipTests: true };

    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const files = tree.files;
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.store.ts`)
    ).toBeGreaterThanOrEqual(0);
    expect(
      files.indexOf(`${projectPath}/src/app/foo/foo.store.spec.ts`)
    ).toEqual(-1);
  });

  it('should register the component store in the provided module', async () => {
    const options = { ...defaultOptions, module: 'app.module.ts' };

    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.module.ts`);

    expect(content).toMatch(/FooStore/);
  });

  it('should register the component store in the provided component', async () => {
    const options = { ...defaultOptions, component: 'app.component.ts' };

    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const content = tree.readContent(`${projectPath}/src/app/app.component.ts`);

    expect(content).toMatch(/FooStore/);
  });

  it('should fail if specified component does not exist', async () => {
    const options = {
      ...defaultOptions,
      component: `${projectPath}/src/app/app.componentXXX.ts`,
    };
    let thrownError: Error | null = null;
    try {
      await schematicRunner.runSchematicAsync(
        'component-store',
        options,
        appTree
      );
    } catch (err) {
      thrownError = err;
    }
    expect(thrownError).toBeDefined();
  });

  it('should inject the component store correctly', async () => {
    const options = { ...defaultOptions };
    const tree = await schematicRunner
      .runSchematicAsync('component-store', options, appTree)
      .toPromise();
    const content = tree.readContent(
      `${projectPath}/src/app/foo/foo.store.spec.ts`
    );

    expect(content).toMatch(/componentStore = new FooStore()/);
  });
});
