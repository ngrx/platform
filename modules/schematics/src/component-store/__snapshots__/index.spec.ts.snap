// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component-store should import into a specified module when the module provided 1`] = `
"import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';
import { FooStore } from './foo/foo.store';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule
  ],
  providers: [FooStore],
  bootstrap: [AppComponent]
})
export class AppModule { }
"
`;

exports[`component-store should inject the component store correctly into the spec 1`] = `
"import { FooStore } from './foo.store';

describe('FooStore', () => {
  const componentStore = new FooStore();

  it('should be created', () => {
    expect(componentStore).toBeTruthy();
  });
});
"
`;

exports[`component-store should not be provided into the module by default 1`] = `
"import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';

import { AppComponent } from './app.component';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
"
`;

exports[`component-store should register the component store in the provided component 1`] = `
"import { Component } from '@angular/core';
import { FooStore } from './foo/foo.store';

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css'],
  providers: [FooStore]
})
export class AppComponent {
  title = 'bar';
}
"
`;
